File: Code/Java/Reversing_linkedlist.java
/*Here Pointer to the head node of a linked list is given, 
 * the task is to reverse the linked list. We are going to reverse the list by changing the links between nodes.We are using Iterative method.*/
/*Initialize three pointers prev as NULL, curr as head and next as NULL.
          Iterate through the linked list. In loop, do following. 
          ->Before changing next of current, 
            -> store next node 
                  next = curr->next
            -> Now change next of current 
           *This is where actual reversing happens 
                  curr->next = prev 
                ->Move prev and curr one step forward. 
                  prev = curr 
                  curr = next*/


import java.util.*;   //used for Scanner function
import java.io.*;     //used for IOException

class Node{                    //creating Node class 
    int data;
    Node next;
    
    Node(int x){
        data = x;
        next = null;
    }
    
}
public class Reversing_linkedlist {


                 static void printList(Node node)   //Function to Print LinkedList
                 { 
	                   while (node != null) 
	                   { 
		                   System.out.print(node.data + " "); 
		                   node = node.next;                         //referring to next node
	                   } 
	                System.out.println(); 
                  } 
         public static void main(String args[]) throws IOException {          //Main starts
    
                        Scanner sc =new Scanner(System.in);                  //Class used for input taking
                        System.out.println("Enter The Numbers Of Nodes in LinkedList:");
    	                int n = sc.nextInt();                                //Number of nodes
    
                        Node head = new Node(sc.nextInt());
                        Node tail = head;
    
                        for(int i=0; i<n-1; i++)                              //For Loop for taking Input
                        {
                             tail.next = new Node(sc.nextInt());
                             tail = tail.next;
                        }
                          
                          System.out.println("Entered LinkedList is:");
                          printList(head);  
                          Solution ob = new Solution();                       // object creation for solution class
                          head = ob.reverseList(head);                         //using ob object reverseList is called
                          System.out.println("Reversed LinkedList is:");
                          printList(head); 
                         
            } 
 }
class Solution{
  
    Node reverseList(Node head)                 //Function for reversing
    {  
    	Node current = head;
        Node prev = null;
        while(current!=null){                       //Checking if the linked list is not null
        head = current.next;                        //assigning next of current to head
        current.next = prev;                        //next of current node = prev;
        prev = current;                             //prev is equal to current node
        current = head;                             //current node is equal to head
       }
    
  return prev;                                      
      
   }
}

/*INPUT AND OUTPUT*/
/*
 * 
 *Enter The Numbers Of Nodes in LinkedList:
  5
  1
  2
  3
  4
  5
  Entered LinkedList is:
  1 2 3 4 5 
  Reversed LinkedList is:
  5 4 3 2 1 
 */
/*
 *Enter The Numbers Of Nodes in LinkedList:
  4
  9
  8
  7
  6
  Entered LinkedList is:
  9 8 7 6 
  Reversed LinkedList is:
  6 7 8 9 
 */
/*TIME COMPLEXITY & SPACE COMPLEXITY*/
/*TIME COMPLEXITY:-O(n)
 * SPACE COMPLEXITY:-O(1)*/

File: Code/Python/Reversing_Linked_list.py
#Python code to Reverse Linked List

# Each element in the Linked List is a Node
# Node consist of two things 1:Data 2:address to next node

#We are using iterative process to reverse a linked list

class Node: #created a class name Node

    #creating a constructor to initialise data and next(pointing to next node)
    #self is keyword in python that has all instance of the class
    def __init__(self,data):
    
        self.data=data #1st field:Data field
        self.next=None #2nd field:Address field
    

class LinkedList: #creating a class where all the required function of linked list will be initialized
    #creating a constructor to initialise head
    def __init__(self):
    
        self.head=None
    
    #function to reverse a linked List
    def reverse(self):
    #we will assign two pointer prev and current
    #prev will point to the current's prev value
    #current will point to current value
        prev=None 
        current=self.head
        while(current is not None):
            next=current.next  #we will assign next pointer to get value of current's next value
            current.next=prev
            prev=current
            current=next
        
        self.head=prev
    #push function to insert new date
    def push(self,new_data):
        new_node=Node(new_data)
        new_node.next=self.head
        self.head=new_node
    
    #print function to print the linked list
    def printLinkedList(self):
        temp=self.head
        while(temp is not None):
            print(temp.data)
            temp=temp.next
            
            
#main code or Driver's Code
create_linkedlist=LinkedList()
item=1
while(item<=5):
    x=int(input("Enter item "+str(item)+" "))
    create_linkedlist.push(x)
    item=item+1
    
print("Given Linked List:")
create_linkedlist.printLinkedList()
create_linkedlist.reverse()
print("Reverse Linked List:")
create_linkedlist.printLinkedList()


#test-case 1: 1 2 3 4 5
#output:Given Linked List:
# 5
# 4
# 3
# 2
# 1
# Reverse Linked List:
# 1
# 2
# 3
# 4
# 5


#test case 2:
# Enter item 1 54
# Enter item 2 22
# Enter item 3 33
# Enter item 4 89
# Enter item 5 100
# Given Linked List:
# 100
# 89
# 33
# 22
# 54
# Reverse Linked List:
# 54
# 22
# 33
# 89
# 100

#time-complexity:O(n)
#space-complexity:O(n)
